mm2src/coins/lightning/ln_utils.rs

```diff
@@ -79,9 +79,17 @@ pub async fn init_db(ctx: &MmArc, ticker: String) -> EnableLightningResult<Sqlit
     Ok(db)
 }
 
-pub fn init_keys_manager(ctx: &MmArc) -> EnableLightningResult<Arc<KeysManager>> {
+pub fn init_keys_manager(platform: &Platform) -> EnableLightningResult<Arc<KeysManager>> {
     // The current time is used to derive random numbers from the seed where required, to ensure all random generation is unique across restarts.
-    let seed: [u8; 32] = ctx.secp256k1_key_pair().private().secret.into();
+    // TODO validate that this is right
+    let seed: [u8; 32] = platform
+        .coin
+        .as_ref()
+        .priv_key_policy
+        .key_pair_or_err()?
+        .private()
+        .secret
+        .into();
```

Comment:
My understanding is that if mm2 is started with a global hd account the key pair here will be the one corresponding to a valid child address so I don't think this will be a problem. Now, this private key will be used as seed in `KeysManager` to generate a new master key that should generate other multiple child keys (a total of six) for node public/private key, closing channels address, and some of these child keys (e.g. channel_master_key) are used as master keys again :) You can check the `KeysManager::new` function internals for the exact details https://github.com/KomodoPlatform/atomicDEX-API/blob/5627cded1616e34c67b795023514a2a86b4c3bba/mm2src/coins/lightning/ln_utils.rs#L95
We have two choices here, either have one lightning node per user that uses the below master seed https://github.com/KomodoPlatform/atomicDEX-API/blob/5627cded1616e34c67b795023514a2a86b4c3bba/mm2src/crypto/src/privkey.rs#L121-L122 as it's `KeysManager` seed, or to give the user the ability to have a lightning wallet/node for every child address using the current implementation in this PR. I think we would need @artemii235 input on which of these options is prefered.

Reply:
@shamardy Will this effect from which address the funding transaction will be sent and to which address closed channel funds will be claimed?

Reply:
You are right. For every key pair there will be a different funding/claiming address (funding and claiming addresses are one and the same address of the platform coin), but currently lightning can't be enabled in `HDWallet` mode, I even added a todo for this in the past, I thought then that HDWallet meant hardware  :) https://github.com/KomodoPlatform/atomicDEX-API/blob/5627cded1616e34c67b795023514a2a86b4c3bba/mm2src/coins/lightning.rs#L1178-L1184
https://github.com/KomodoPlatform/atomicDEX-API/blob/5627cded1616e34c67b795023514a2a86b4c3bba/mm2src/coins/lightning/ln_events.rs#L528-L535
https://github.com/KomodoPlatform/atomicDEX-API/blob/5627cded1616e34c67b795023514a2a86b4c3bba/mm2src/coins/rpc_command/lightning/open_channel.rs#L148
If we agreed to have a lightning wallet/node for every child address, I should add HDWallet implementation for lightning to this checklist https://github.com/KomodoPlatform/atomicDEX-API/issues/1045#issue-971642359

Reply:
One more note, we can always implement our own lightning `KeysManager` if we needed that. The only requirement for `KeysManager` is to implement the [`KeysInterface` trait](https://github.com/lightningdevkit/rust-lightning/blob/e55e0d53c74ba8c810ad36ea39e0ea1621873c34/lightning/src/chain/keysinterface.rs#L404). But I think the one implemented in rust-lightning is very good for preserving privacy, we break this privacy in our lightning implementation when claiming the closing funds back to the same address the channel was opened from. I guess to continue the preserving of privacy with HDWallets we can let the user decide which child address to generate funding from and which address to claim the closing funds to for every channel, if we choose to use only one lightning node for all platform coin addresses then we would need to implement our own lightning `KeysManager` where there is one node address using the master seed and change the destination script, the shutdown pubkey when switching addresses.

Reply:
`HDWallet` is related to Hardware wallet only.
`Global HD account` is implemented the way that if the user wants to switch its Global HD account, he needs to restart mm2 instance with another `conf.hd_account_id` index.
So for a running mm2 instance with an `conf.hd_account_id` value there is only one address (therefore one key-pair) for any coin.

Reply:
mmm, I understand now but I think we should change `DerivationMethod::HDWallet` naming to convey that it's for hardware wallets only to avoid confusion.
So, in light of this info, for every `conf.hd_account_id` there will be a different lightning address/node with different channels backups etc..., I still see no problem caused in lightning by the current implementation :)

Reply:
For the privacy issues I detailed here https://github.com/KomodoPlatform/atomicDEX-API/pull/1512#discussion_r1014059513, there will be a different destination script and shutdown pubkey also for every `hd_account_id`, I guess we can add an option for the user to claim the closing transaction of a channel to a different `hd_account_id` or a different address of his choosing making lightning channels opening/closing completely private. Of Course this is not a priority now, but just adding my thoughts about HD accounts implications for lightning.

Reply:
Ok thank you! `DerivationMethod::HDWallet` is really a confusing name, because it overlaps with `Global HD account`. But they are both related to HD wallet:
1) Global HD account: `one coin -> one address`.
2) DerivationMethod::HDWallet: `one coin -> multiple accounts`, `account -> multiple addresses (internals and/or externals)`.

I explained it in more details here: https://github.com/KomodoPlatform/air_dex/issues/502#issuecomment-1193167304

`DerivationMethod::HDWallet` is currently used with Trezor only, but it's designed the way that it's independent from the private key policy. Thus later we can add new `PrivKeyPolicy::CoinHDAccount`, for example. The only difference from `PrivKeyPolicy::CoinHDAccount` will be how transactions are signed.

If we're sure that we don't plan to support such derivation in the future, I can rename `DerivationMethod::HDWallet` to `DerivationMethod::HardwareWallet`.
@artemii235 could you please advice?

Reply:
> If we're sure that we don't plan to support such derivation in the future

We can't 100% sure, I guess 🙂 Though, I would prefer the naming to reflect the current meaning of the code. We can figure out one more renaming later if required.